File: desktop/tests/Qlipq.Ffmpeg.Tests/EstimateTests.cs
using Qlipq.Core;
using Qlipq.Ffmpeg;
using Xunit;

namespace Qlipq.Ffmpeg.Tests;

public class EstimateTests
{
    private static readonly MediaInfo Media = new()
    {
        DurationSec = 100,
        Width = 1920,
        Height = 1080,
        VideoCodec = "h264",
        Fps = 60,
        AudioStreams = [],
        SizeBytes = 1_000_000_000, // 1 GB over 100s
    };

    private static EditSpec OneAudio() => new() { AudioTracks = [new AudioTrackSpec { Index = 0, Enabled = true, Volume = 1 }] };

    private static ResolvedEncode Copy() =>
        new() { Video = new VideoEncodeOptions(), Audio = new AudioEncodeOptions { Bitrate = "192k" }, Reencode = false };

    private static ResolvedEncode Crf(Action<VideoEncodeOptions>? over = null)
    {
        var video = new VideoEncodeOptions { Codec = "libx264", Crf = 23 };
        over?.Invoke(video);
        return new ResolvedEncode { Video = video, Audio = new AudioEncodeOptions { Bitrate = "192k" }, Reencode = true };
    }

    [Fact]
    public void StreamCopyScalesSourceSizeByKeptDuration()
    {
        var full = Estimate.EstimateExportSize(Media, OneAudio(), Copy());
        Assert.True(Math.Abs(full.Bytes - 1_000_000_000) < 500);
        Assert.False(full.Approximate);

        var half = Estimate.EstimateExportSize(
            Media,
            new EditSpec { AudioTracks = OneAudio().AudioTracks, Trim = new TrimSpec { StartSec = 0, EndSec = 50 } },
            Copy());
        Assert.True(Math.Abs(half.Bytes - 500_000_000) < 500);
    }

    [Fact]
    public void BitrateModeIsExactBitrateTimesDuration()
    {
        var enc = new ResolvedEncode
        {
            Video = new VideoEncodeOptions { BitrateKbps = 8000 },
            Audio = new AudioEncodeOptions { Bitrate = "0k" },
            Reencode = true,
        };
        var r = Estimate.EstimateExportSize(Media, new EditSpec { AudioTracks = [] }, enc);
        Assert.True(Math.Abs(r.Bytes - 8000.0 * 1000 * 100 / 8) < 500);
        Assert.False(r.Approximate);
    }

    [Fact]
    public void CrfEstimateIsApproximateAndMonotonic()
    {
        var better = Estimate.EstimateExportSize(Media, OneAudio(), Crf(v => v.Crf = 18));
        var worse = Estimate.EstimateExportSize(Media, OneAudio(), Crf(v => v.Crf = 28));
        Assert.True(better.Approximate);
        Assert.True(better.Bytes > worse.Bytes);

        var downscaled = Estimate.EstimateExportSize(Media, OneAudio(), Crf(v => { v.Crf = 23; v.ScaleHeight = 540; }));
        var fullRes = Estimate.EstimateExportSize(Media, OneAudio(), Crf(v => v.Crf = 23));
        Assert.True(downscaled.Bytes < fullRes.Bytes);
    }

    [Fact]
    public void H265EstimatesSmallerThanH264AtSameCrf()
    {
        var h264 = Estimate.EstimateExportSize(Media, OneAudio(), Crf(v => v.Codec = "libx264"));
        var h265 = Estimate.EstimateExportSize(Media, OneAudio(), Crf(v => v.Codec = "libx265"));
        Assert.True(h265.Bytes < h264.Bytes);
    }

    [Fact]
    public void ZeroLengthOutputEstimatesToZero()
    {
        var r = Estimate.EstimateExportSize(
            Media,
            new EditSpec { AudioTracks = OneAudio().AudioTracks, Trim = new TrimSpec { StartSec = 10, EndSec = 10 } },
            Copy());
        Assert.Equal(0, r.Bytes);
    }
}

File: iced/crates/qlipq-ffmpeg/tests/estimate_tests.rs
use qlipq_core::edit_spec::{AudioTrackSpec, EditSpec, TrimSpec};
use qlipq_core::media::MediaInfo;
use qlipq_ffmpeg::args::{AudioEncodeOptions, ResolvedEncode, VideoEncodeOptions};
use qlipq_ffmpeg::estimate::*;

fn media() -> MediaInfo {
    MediaInfo {
        duration_sec: 100.0,
        width: 1920,
        height: 1080,
        video_codec: "h264".into(),
        fps: 60.0,
        audio_streams: vec![],
        size_bytes: Some(1_000_000_000),
    }
}

fn one_audio() -> EditSpec {
    EditSpec { trim: None, crop: None, audio_tracks: vec![AudioTrackSpec { index: 0, enabled: true, volume: 1.0 }] }
}

fn copy() -> ResolvedEncode {
    ResolvedEncode {
        video: VideoEncodeOptions::default(),
        audio: AudioEncodeOptions { codec: None, bitrate: Some("192k".into()) },
        reencode: false,
    }
}

fn crf(configure: impl FnOnce(&mut VideoEncodeOptions)) -> ResolvedEncode {
    let mut video = VideoEncodeOptions { codec: Some("libx264".into()), crf: Some(23), ..Default::default() };
    configure(&mut video);
    ResolvedEncode { video, audio: AudioEncodeOptions { codec: None, bitrate: Some("192k".into()) }, reencode: true }
}

#[test]
fn stream_copy_scales_by_kept_duration() {
    let full = estimate_export_size(&media(), &one_audio(), &copy());
    assert!((full.bytes - 1_000_000_000.0).abs() < 500.0);
    assert!(!full.approximate);

    let half = estimate_export_size(
        &media(),
        &EditSpec { trim: Some(TrimSpec { start_sec: 0.0, end_sec: 50.0 }), crop: None, audio_tracks: one_audio().audio_tracks },
        &copy(),
    );
    assert!((half.bytes - 500_000_000.0).abs() < 500.0);
}

#[test]
fn bitrate_mode_is_exact() {
    let enc = ResolvedEncode {
        video: VideoEncodeOptions { bitrate_kbps: Some(8000), ..Default::default() },
        audio: AudioEncodeOptions { codec: None, bitrate: Some("0k".into()) },
        reencode: true,
    };
    let r = estimate_export_size(&media(), &EditSpec { trim: None, crop: None, audio_tracks: vec![] }, &enc);
    assert!((r.bytes - 8000.0 * 1000.0 * 100.0 / 8.0).abs() < 500.0);
    assert!(!r.approximate);
}

#[test]
fn crf_estimate_is_approximate_and_monotonic() {
    let better = estimate_export_size(&media(), &one_audio(), &crf(|v| v.crf = Some(18)));
    let worse = estimate_export_size(&media(), &one_audio(), &crf(|v| v.crf = Some(28)));
    assert!(better.approximate);
    assert!(better.bytes > worse.bytes);

    let downscaled = estimate_export_size(&media(), &one_audio(), &crf(|v| { v.crf = Some(23); v.scale_height = Some(540); }));
    let full_res = estimate_export_size(&media(), &one_audio(), &crf(|v| v.crf = Some(23)));
    assert!(downscaled.bytes < full_res.bytes);
}

#[test]
fn h265_smaller_than_h264() {
    let h264 = estimate_export_size(&media(), &one_audio(), &crf(|v| v.codec = Some("libx264".into())));
    let h265 = estimate_export_size(&media(), &one_audio(), &crf(|v| v.codec = Some("libx265".into())));
    assert!(h265.bytes < h264.bytes);
}

#[test]
fn zero_length_estimates_zero() {
    let r = estimate_export_size(
        &media(),
        &EditSpec { trim: Some(TrimSpec { start_sec: 10.0, end_sec: 10.0 }), crop: None, audio_tracks: one_audio().audio_tracks },
        &copy(),
    );
    assert_eq!(r.bytes, 0.0);
}
